### ISSUE TITLE
DetailsPage' pages do not support dynamic components with state

### ISSUE THREAD

**Issue by suomiy** [openshift/console#1567]
`PageHeading` and `HorizontalNav`are getting new data from Firehose. 
`HorizontalNav` itself has `Switch` with `Route`s inside. The router does not support dynamically creating pages with different props with full support of lifecycle methods. Meaning the component is unmounted and mounted again with each underlying resource change.

Real example:
Any change of state is discarded, if any change occurs to, for example the VM resource, while it is being edited inside the Overview page.

Proposed Solution. Move react Router to the top and do not pass dynamic props to it (just namespace, name, etc.):

```
DetailsPage
  Firehose
    PageHeading
  HorizontalNav
    Switch
      Route
        Firehose
          PageComponent
```

We could also just watch the resource (`watchK8sObject`) in `DetailsPage` and pass reduxIds down to the `PageHeading` and `PageComponent`

**Comment by spadgett** [comment]
Related to #1558 although it doesn't sound like #1558 will fix this specific problem from the description.

cc @christianvogt

**Comment by christianvogt** [comment]
@suomiy i'm not sure i'm following you. Can you give a specific example of steps and how the problem manifests itself because I do not see the unmounting of components.

**Comment by suomiy** [comment]
Please read the documentation of Route https://reacttraining.com/react-router/web/api/Route/route-render-methods. 

We are using inline rendering of Route component (in HorizontalNav) which means that React lifecycle methods are not properly called - so it is not easy to debug.  

Only noticable change is that the state is lost. I am not aware of any tab with state so you could try to experiment with edit option of VMs in https://github.com/kubevirt/web-ui.

Or you can try to pass component to the Route directly instead of render function and you should see the lifecycle methods called properly - with undesirable unmounts happening with every resource change. E.g. just look at pod details tab and edit pod in another browser window.

**Comment by spadgett** [comment]
Secret -> Reveal Values could be a good page to test. Revealed/hidden is tracked using component state.

**Comment by christianvogt** [comment]
@suomiy I think you are mistaken in that an inline function to the `render` prop does not mean it will unmount the child components. That only occurs with the `component` prop if an inline function is used.

Horizontal nav uses an inline `render` prop yes. But this will only cause a component to unmount if the return value of the `render` function is a different component. If you look at the docs, or the source code of `Route`, it shows that `component` prop uses `React.createElement` which will cause the old one to unmount.

**Comment by christianvogt** [comment]
@spadgett @suomiy Secrets works for me without losing state. Edit an annotation from the secrets page and watch the update of the page. The revealed secret is still revealed.

**Comment by spadgett** [comment]
This should in fact be fixed by #1558. If you still see the problem, let us know.

/close

**Comment by suomiy** [comment]
I am sorry for the unnecessary alert. I cannot reproduce this either (on my env).

It only happens in one of our environments for running the tests (quite consistently though). I was debugging it there and this seemed like the most probable cause, because it exhibited all the symptoms described here.
